File: myblog-frontend/src/services/post.ts
import { Post, PostDAO } from './../models/Post';
import axios from './axios'

const apiUrl = '/posts'

const responseToPost = (p: Post) => ({
  ...p,
  created_at: new Date(p.created_at)
})
export class PostService{
  static async create(data: PostDAO): Promise<Post>{
    try {
      const post = (await axios.post<Post>(apiUrl, data)).data
      return responseToPost(post)
    } catch (error: any) {
      throw new Error(error.response.data.error)
    }
  }
  
  static async getPosts(): Promise<Post[]>{
    return (await axios.get<Post[]>(apiUrl)).data
      .map(responseToPost)
  }
  
  static async getById(id: number): Promise<Post>{
    try {
      const post = (await axios.get<Post>(`${apiUrl}/${id}`)).data
      return responseToPost(post)
    } catch (error: any) {
      throw new Error(error.response.data.error)
    }
  }
  static async delete(id: number) {
    try {
      await axios.delete(`${apiUrl}/${id}`)
    } catch (error: any) {
      throw new Error(error.response.data.error)
    }
  }
}

File: myblog-backend/repository/post.go
package repository

import (
	"github.com/baiest/myblog/myblog-backend/db"
	"github.com/baiest/myblog/myblog-backend/models"
)

type IPostRepository interface {
	Create(data models.PostCreate) (*models.Post, error)
	GetAll() []models.Post
	GetById(id models.IdPost) (*models.Post, error)
	// Update()
	Delete(id models.IdPost) error
}

type PostRepository struct{}

func (p PostRepository) Create(data models.PostCreate) (*models.Post, error) {
	post := models.NewPost(data)
	created := db.DB.Create(&post)
	if err := created.Error; err != nil {
		return nil, err
	}

	return post, nil
}

func (p PostRepository) GetAll() []models.Post {
	var posts []models.Post
	db.DB.Preload("User").Find(&posts)
	return posts
}

func (p PostRepository) GetById(id models.IdPost) (*models.Post, error) {
	post := models.Post{
		Id: id,
	}
	found := db.DB.Preload("User").First(&post)
	if err := found.Error; err != nil {
		return nil, err
	}
	return &post, nil
}

func (p PostRepository) Delete(id models.IdPost) error {
	return db.DB.Delete(models.Post{}, id).Error
}
